registry/virtual.go

```diff
@@ -1,50 +1,455 @@
+	}
+	defer func() {
+		cerr := file.Close()
+		if err == nil {
+			err = cerr
+		}
+	}()
+
+	stats, err := file.Stat()
+	if err != nil {
+		return "", 0, err
+	}
+	length = stats.Size()
+
+	hasher := sha256.New()
+	if _, err := io.Copy(hasher, file); err != nil {
+		return "", 0, err
+	}
+	h := hasher.Sum(nil)
+	hash = hex.EncodeToString(h)
+
+	prefix := base.Prefix(s.Name)
+	if _, err := file.Seek(0, io.SeekStart); err != nil {
+		return "", 0, err
+	}
+
+	if err = base.Storage.EnsureExists(prefix); err != nil {
+		return "", 0, err
+	}
+	if err = base.Storage.Create(prefix, hash, "application/gzip", file); err != nil {
+		return "", 0, err
+	}
+
+	return hash, length, nil
+}
+
+func deleteOverwrittenTarball(s base.VirtualSpace, version *Version) error {
+	h, ok := version.AttachmentReferences["tarball"]
+	if ok {
+		prefix := base.Prefix(s.Name)
+		if err := base.Storage.Remove(prefix, h); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func getOriginalTarball(space *space.Space, version *Version) (io.Reader, error) {
+	url, err := url.Parse(version.URL)
+	if err != nil {
+		return nil, err
+	}
+	filename := filepath.Base(url.Path)
+
+	att, err := FindVersionAttachment(space, version, filename)
+	if err != nil {
+		return nil, err
+	}
 
+	return att.Content, nil
+}
+
+func generateOverwrittenTarball(version *Version, overwrite map[string]interface{}, input io.Reader) (file string, manifest map[string]interface{}, icon string, err error) {
+	var newManifest map[string]interface{}
+
+	var originManifest Manifest
+	if err := json.Unmarshal(version.Manifest, &originManifest); err != nil {
+		return "", nil, "", err
+	}
+
+	iconFilename := originManifest.Icon
+	manifestFilename := "manifest." + version.Type
+
+	iconChecksum, iconOverwritten := overwrite["icon"].(string)
+	var iconContent *bytes.Buffer
+	if iconOverwritten {
+		iconContent, _, err = base.GlobalAssetStore.Get(iconChecksum)
+		if err != nil {
+			return "", nil, "", err
+		}
+	}
+	name, nameOverwritten := overwrite["name"].(string)
+
+	inputGzip, err := gzip.NewReader(input)
+	if err != nil {
+		return "", nil, "", err
+	}
+	defer inputGzip.Close()
+	inputTar := tar.NewReader(inputGzip)
+
+	prefix := fmt.Sprintf("%s_%s_*.tar.gz", version.Slug, version.Version)
+	outputFile, err := ioutil.TempFile("", prefix)
+	if err != nil {
+		return "", nil, "", err
+	}
+	file = outputFile.Name()
+	// Tricky return case from here. Never `return nil, err`.
+	// File is already created on filesystem, so we need to return it in all case to be able to clean it on the caller.
+	// We can'm `defer os.Remove` here, because caller need the file for storage…
+
+	outputGzip := gzip.NewWriter(outputFile)
+	if err != nil {
+		return "", nil, "", err
+	}
+	defer func() {
+		cerr := outputGzip.Close()
+		if err == nil {
+			err = cerr
+		}
+	}()
+
+	outputTar := tar.NewWriter(outputGzip)
+	defer func() {
+		cerr := outputTar.Close()
+		if err == nil {
+			err = cerr
+		}
+	}()
+
+out:
```

Comment:
Please, no label breaks!

Reply:
Do you see a way to avoid that here?
Have to break out the for but already in a switch :(

Reply:
As there is no logic after the for loop, we can just `return`